CHANGELOG.md

```diff
@@ -19,6 +19,47 @@ to ensure the changelog can be parsed correctly.
 
 -------------------------------------------------------------------------------
 
+## 2023-04-05: 0.23
+
+spicyjpeg:
+
+- libc: Fixed bugs in some string manipulation and number parsing functions.
+  Added `memccpy()`. Removed the `SYSTEM.CNF` argument parser as it would
+  interfere with manual argc/argv passing in some edge cases.
+
+- psxgpu: Added `IsIdleGPU()` and fixed a bug in `SetVideoMode()`. Added new
+  "tagless" primitive structures with the `_T` suffix (e.g. `POLY_FT4_T`,
+  `SPRT_T`, ...) and related macros (`setPolyFT4_T()`, `setSPRT_T()`, ...) that
+  can be used to group multiple primitives into a single display list packet
+  for better performance. Fixed some macros not working properly with negative
+  values. Renamed `DR_MASK` and `setDrawMask()` to `DR_STP` and `setDrawStp()`
+  respectively for consistency with the official SDK.
+
+- psxcd: Added `CdUnlock()`. Improved reliability of `CdGetRegion()` on early
+  Japanese console models.
+
+- psxetc: Added `SetDMAPriority()` and `GetDMAPriority()`.
+
+- psxapi: Added PCDRV (host file I/O) API, declared in the `psxsn.h` header.
+
+- psxpress: Fixed bugs in the MDEC API. Added support for bitstream version 3
+  to the GTE-accelerated decoder. Replaced `DecDCTvlcCopyTable()` with
+  `DecDCTvlcCopyTableV2()` and `DecDCTvlcCopyTableV3()` for better control over
+  how much of the scratchpad is used for lookup tables.
+
+- examples: `mdec/strvideo` is now (finally) functional on real hardware and
+  can properly detect the end of a video file even if immediately followed by
+  another video on the disc. Added previously missing CD image dependencies to
+  the CMake scripts of the examples that make use of CD images.
+
+- tools: `elf2x` no longer adds region strings to converted executables, fixing
+  issues with DuckStation's region autodetection. Updated `mkpsxiso` to 2.03 in
+  order to fix a bug with its thread pool implementation on Linux.
+
+- Updated binutils and recompiled the GCC toolchain with the
+  `--disable-hosted-libstdcxx` option, allowing a subset of the C++ STL to be
+  used in PSn00bSDK projects.
```

Comment:
This is really exciting! Back when I wrote the _unmoving_ lib, I had to stick to bare-metal C++20. Managed to get it working with the SDK eventually, but I found it was quite tricky to pin down the exact configure options needed to get it to build even the baremetal C++ stdlib which is required for many C++20 features that are meant to be implemented as builtins.

Can you refer me to any demos or info on what parts of the STL are available with this technique?

Reply:
The `--disable-hosted-libcxx` option is a fairly recent addition to GCC; or rather, it used to be broken and was only fixed relatively recently (and only because Nicolas Noble of pcsx-redux fame pointed it out). Currently I have no code examples showcasing usage of this subset of the STL, but the following headers are already available in v0.23 preview builds:

- `atomic`
- `bit`
- `cfloat`
- `ciso646`
- `climits`
- `compare`
- `concepts`
- `coroutine`
- `cstdalign`
- `cstdarg`
- `cstdbool`
- `cstddef`
- `cstdint`
- `cstdlib`
- `cxxabi.h`
- `exception`
- `initializer_list`
- `limits`
- `new`
- `source_location`
- `type_traits`
- `typeinfo`
- `version`

If you wish to contribute by adding more examples or updating the existing C++ example (which desperately needs a rewrite), feel free to go ahead and open more PRs. I'll review everything once I am ready to release v0.23.

Reply:
This is _great_! I'm so happy to see `<compare>` in there in particular, along with a bunch of other quality-of-life C++ stuff that's meant to be available in baremetal. Hooray for Nicolas Noble!

I will deffo check this out when I get the time, maybe I will even be able to rewrite that C++20 lib I wrote for PSX to use more of these quality-of-life features now that the toolchain supports them.

It wouldn't surprise me if that broken option that you mentioned was why I couldn't get it completely working in a convenient way on my end.

Reply:
I'll deffo let you know if I find anything useful or interesting to contribute back as a C++ example, I will certainly be able to contribute instructions for the C++20 compiler if needed.

On the topic of this, it'd be great if we could produce some unit testing of the SDK libraries. I realise that unit-testing on a limited baremetal platform like PSX is a bit difficult, but I'm wondering of the possibility of a series of test cases which get built as a number of separate binaries that could then be ran under PCSXR or something. Not sure about how to get test case feedback out in an autonomously-verifiable way, though!

Reply:
It's funny you mention unit testing, because Nicolas has been working on that too (as well as [his own PS1 SDK with a full C++ STL](https://github.com/grumpycoders/pcsx-redux/tree/main/src/mips/psyqo), which I should have mentioned before). Performing automated testing on pcsx-redux should be simple enough given that it has proper GDB support; doing it on real hardware isn't too far fetched either, considering that Unirom is getting support for remote debugging through the console's serial port. Of course this would need some additional infrastructure such as a way to script the emulator (in fact pcsx-redux already exposes various bits of functionality via Lua and HTTP APIs), but it is still entirely within the realm of possibility.

If you want to discuss this further and get some feedback, you can reach out to me, Nicolas as well as other emulator developers on the [PSX.Dev Discord server](https://discord.gg/QByKPpH).